servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java

```diff
@@ -978,6 +981,83 @@ boolean shouldRetryOperation(ResourceException ex, int retryCount) {
         return retry;
     }
 
+    public void executePutPolicyVersion(ResourceContext ctx, String domainName, String policyName, String version, String fromVersion,
+                                        String auditRef, String caller) {
+        // our exception handling code does the check for retry count
+        // and throws the exception it had received when the retry
+        // count reaches 0
+
+        for (int retryCount = defaultRetryCount; ; retryCount--) {
+
+            try (ObjectStoreConnection con = store.getConnection(false, true)) {
+
+                // first verify that auditing requirements are met
+
+                checkDomainAuditEnabled(con, domainName, auditRef, caller, getPrincipalName(ctx), AUDIT_TYPE_POLICY);
+
+                // verify the policy hasn't reached maximum number of versions allowed
+
+                List<String> policyVersions = con.listPolicyVersions(domainName, policyName);
+                if (policyVersions.size() >= maxPolicyVersions) {
+                    con.rollbackChanges();
+                    throw ZMSUtils.quotaLimitError("unable to put policy: " + policyName + ", version: " + version + ", max number of versions reached (" + maxPolicyVersions + ")", caller);
+                }
+
+                // retrieve our source policy version
+
+                Policy originalPolicy = getPolicy(con, domainName, policyName, fromVersion);
+
+                // check that quota is not exceeded
+
+                quotaCheck.checkPolicyQuota(con, domainName, originalPolicy, caller);
+
+                // now process the request
+
+                StringBuilder auditDetails = new StringBuilder(ZMSConsts.STRING_BLDR_SIZE_DEFAULT);
+                if (!con.insertPolicyVersion(domainName, version, originalPolicy)) {
+                    con.rollbackChanges();
+                    throw ZMSUtils.internalServerError("unable to put policy: " + originalPolicy.getName() + ", version: " + version, caller);
+                }
+
+                // open our audit record
+
+                auditDetails.append("{\"name\": \"").append(policyName);
+                auditDetails.append("\", \"version\": \"").append(version);
+
+                // now we need process our policy assertions
+
+                List<Assertion> newAssertions = originalPolicy.getAssertions();
+                if (newAssertions != null) {
+                    for (Assertion assertion : newAssertions) {
```

Comment:
for completeness sake we must copy all the assertion conditions as well when we're making a copy of the policy. at some point we might use this functionality in microsegmentation as well so we need to make sure all conditions are available as well.

Reply:
fixed